cg_openmm/simulation/rep_exch.py

```diff
@@ -109,35 +105,56 @@ def process_replica_exchange_data(
     n_particles = np.shape(reporter.read_sampler_states(iteration=0)[0].positions)[0]
     temps = np.array([temp._value for temp in temperature_list])
     beta_k = 1 / (kB * temps)
-    for k in range(len(replica_energies)):
+    n_replicas = len(temperature_list)
+    for k in range(n_replicas):
         replica_energies[:, k, :] *= beta_k[k] ** (-1)
 
     total_steps = len(replica_energies[0][0])
-    replica_positions = unit.Quantity(
-        np.zeros([len(temperature_list), total_steps, n_particles, 3]), unit.nanometer
-    )
+    state_energies = np.zeros([n_replicas, total_steps])
+
+    # there must be some better way to do this as list comprehension.
+    for step in range(total_steps):
+        for state in range(n_replicas):
+            state_energies[state, step] = replica_energies[
+                np.where(replica_state_indices[:, step] == state)[0], 0, step
```

Comment:
Still trying to understand the data extraction from replica_energies 3d array. For a given 'step' (third index of replica_energies), there is a matrix of energies of shape [n_replicas, n_states], but only 'n_states' of those values actually correspond to that iteration, so that's why we need to search with np.where?

I tested this code out and the mean energies and variance seem right. So why is it that we use 0 for all state in the for loop here?

Reply:
Great question. The replica_energies array stores the energy of each sample for each replica, and then each state.  So [0,4,:] are the energies of the 0th replica, in the 4th thermodynamic state.  Since our thermodynamic states are just different temperatures, then all of the energies are the same - so we could use 0, or any other value of the the state index. The np.where(replica_state_indices[:, step] == state) is just asking "Which of the replicas is in the first state?".  The [0] is because the answer is an array format, we want the value out.

Does that make sense?

Reply:
The np.where part I get. But I just tried changing the second index to 1, processing the same .nc file as before, and all of the energies are different. My understanding is that the 0th replica would be in the 4th state for only a small subset of the iterations?

Reply:
That's interesting.  I didn't test that before, but just now I tried switching:

`np.where(replica_state_indices[:,` step] == state)[0], 0, step`

to 

`np.where(replica_state_indices[:, step] == state)[0], 1, step`

and 

`np.where(replica_state_indices[:, step] == state)[0], 5, step`

And got the exact same average energies and variances in all three cases.

Reply:
You're right, I must have missed something going through line by line of the process replica exchange function. The energies are the same regardless of the second index used.

Reply:
Sound good, let me know if there are any other issues before merging.